finality-aleph/src/testing/data_store.rs

```diff
@@ -0,0 +1,645 @@
+        .await;
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn unimported_hopeless_forks_go_through() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10)
+        .await;
+    let forking_block = &blocks[MAX_DATA_BRANCH_LEN + 2];
+    let forks = test_handler
+        .build_branch_upon(&forking_block.hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    for i in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = forks[0..(i as usize)].to_vec();
+        let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+        test_handler.send_data(test_data.clone());
+    }
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.import_branch(blocks.clone(), false).await;
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 1].hash());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN * 2 + 1].hash());
+
+    for _ in 1..=MAX_DATA_BRANCH_LEN {
+        let _message = test_handler
+            .assert_message_out("Did not receive message from Data Store")
+            .await;
+    }
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn imported_hopeless_forks_go_through() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10)
+        .await;
+    let forking_block = &blocks[MAX_DATA_BRANCH_LEN + 2];
+    let forks = test_handler
+        .build_branch_upon(&forking_block.hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    test_handler.import_branch(blocks.clone(), false).await;
+    test_handler.import_branch(forks.clone(), false).await;
+
+    for i in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = forks[0..(i as usize)].to_vec();
+        let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+        test_handler.send_data(test_data.clone());
+    }
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 1].hash());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
```

Comment:
Same as above, the situation is different so please make the errors differ too.

Reply:
I changed the error messages a bit to better reflect the situation, but still they are there same because it's still the same reason why they should not be let through in both cases.